File: src/main/java/org/team2471/frc/lib/motion_profiling/Autonomi.java
package org.team2471.frc.lib.motion_profiling;

import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.Moshi;
import edu.wpi.first.networktables.NetworkTable;
import edu.wpi.first.networktables.NetworkTableEntry;
import edu.wpi.first.networktables.NetworkTableInstance;
import org.team2471.frc.lib.motion_profiling.following.DrivetrainParameters;
import org.team2471.frc.lib.motion_profiling.following.RobotParameters;

import java.util.LinkedHashMap;
import java.util.Map;

public class Autonomi {
    public RobotParameters robotParameters;
    public DrivetrainParameters drivetrainParameters;

    public Map<String, Autonomous> mapAutonomous = new LinkedHashMap<>();

    private static JsonAdapter<Autonomi> jsonAdapter = new Moshi.Builder()
            .add(DrivetrainParameters.getMoshiAdapter())
            .build()
            .adapter(Autonomi.class).indent("\t");

    public Autonomous get(String name) {
        return mapAutonomous.get(name);
    }

    public void put(Autonomous autonomous) {
        mapAutonomous.put(autonomous.name, autonomous);
    }

    public Path2D getPath(String autoName, String pathName) {
        Autonomous autonomous = get(autoName);
        return autonomous.get(pathName);
    }

    public String toJsonString() {
        return jsonAdapter.toJson(this);
    }

    static public Autonomi fromJsonString(String json) {
        Autonomi autonomi;
        try {
            autonomi = jsonAdapter.fromJson(json);
        } catch (Exception e) {
            System.out.println("Constructing Autonomi class from json failed.");
            return null;
        }

        if (autonomi != null)
            autonomi.fixUpTailAndPrevPointers();

        return autonomi;
    }

    private void fixUpTailAndPrevPointers() {
        for (Map.Entry<String, Autonomous> entry : mapAutonomous.entrySet()) {
            entry.getValue().fixUpTailAndPrevPointers();
        }
    }

    public void publishToNetworkTables(NetworkTableInstance networkTableInstance) {
        String json = toJsonString();
        NetworkTable table = networkTableInstance.getTable("PathVisualizer");
        NetworkTableEntry entry = table.getEntry("Autonomi");
        entry.setString(json);
    }
}

File: src/main/java/org/team2471/frc/lib/motion_profiling/Autonomi.kt
package org.team2471.frc.lib.motion_profiling

import org.team2471.frc.lib.motion_profiling.following.DrivetrainParameters.Companion.moshiAdapter
import org.team2471.frc.lib.motion_profiling.following.RobotParameters
import org.team2471.frc.lib.motion_profiling.following.DrivetrainParameters
import org.team2471.frc.lib.motion_profiling.following.SwerveParameters
import org.team2471.frc.lib.motion_profiling.Autonomous
import org.team2471.frc.lib.motion_profiling.Autonomi
import edu.wpi.first.networktables.NetworkTableInstance
import edu.wpi.first.networktables.NetworkTableEntry
import com.squareup.moshi.Moshi
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
import java.lang.Exception
import java.util.LinkedHashMap

class Autonomi {
    var robotParameters = RobotParameters(0.0, 0.0)
    var drivetrainParameters: DrivetrainParameters = SwerveParameters(0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, false)
    var mapAutonomous: MutableMap<String, Autonomous> = LinkedHashMap()
    operator fun get(name: String): Autonomous? {
        return mapAutonomous[name]
    }

    fun put(autonomous: Autonomous) {
        mapAutonomous[autonomous.name] = autonomous
    }

    fun getPath(autoName: String, pathName: String?): Path2D {
        val autonomous = get(autoName)
        return autonomous!![pathName]
    }

    fun toJsonString(): String {
        return jsonAdapter.toJson(this)
    }

    private fun fixUpTailAndPrevPointers() {
        for ((_, value) in mapAutonomous) {
            value.fixUpTailAndPrevPointers()
        }
    }

    fun publishToNetworkTables(networkTableInstance: NetworkTableInstance) {
        val json = toJsonString()
        val table = networkTableInstance.getTable("PathVisualizer")
        val entry = table.getEntry("Autonomi")
        entry.setString(json)
    }

    fun readFromNetworkTables(networkTableInstance: NetworkTableInstance): String {
        val table = networkTableInstance.getTable("PathVisualizer")
        val entry = table.getEntry("Autonomi")
        println(entry.getString("default"))
        return entry.getString("")
    }

    companion object {
        private val jsonAdapter = Moshi.Builder()
            .addLast(KotlinJsonAdapterFactory())
            .add(moshiAdapter)
            .build()
            .adapter(Autonomi::class.java).indent("\t")

        fun fromJsonString(json: String?): Autonomi? {
            val autonomi: Autonomi? = try {
                jsonAdapter.fromJson(json)
            } catch (e: Exception) {
                println("Constructing Autonomi class from json failed.")
                return null
            }
            autonomi?.fixUpTailAndPrevPointers()
            return autonomi
        }
    }
}